content/2020-04-23-value-range-analysis.rst

```diff
@@ -126,6 +126,59 @@ Let's illustrate that analysis through an example:
             print("ok")
         return a, b, c
 
+It's a control flow analysis, so it follows the control flow starting from the
+function entry point. It first meets an ``assert``, so we register that ``1 <= a <= inf``
+(remember, we only use intervals). Then, there's an assignment of constant
+value, so we have:
+
+.. code::
+
+    1 <= a <= inf
+    10 <= b <= 10
+    10 <= c <= 10``
+
+Then comes a ``while`` statement. A while usually has two successors: its body,
+and the next statement. But in that case we evaluate the condition and see that
+it always holds, because we have ``1 <= a``, so we first perform a first round
+of the body, getting, through the two accumulation (let's drop ``c`` for the
+sake of clarity):
+
+.. code::
+
+    0 <= a <= inf
+    11 <= b <=11
+
+Then we're back to the test. The condition no longer always hold, so we need to
+make a decision! The idea here is to perform a *widening*, so we record current
+state, and perform another round, getting ``-1 <= a <= inf; 12 <= b <= 12``.
+Through the comparison of the two states, we can see the evolution of ``a``
+(it shrinks towards ``-inf``) and ``b`` (it grows toward ``+inf``).
+This maybe not super accurate, but it's a correct overestimate.
+So we decide that right before the test, we have:
+
+.. code::
+
+    -inf <= a <= inf
+    11 <= b <= inf
+
+It's safe to apply the condition at the entry point too, so let's constraint our
```

Comment:
constrain

Reply:
Also, I'm not sure why it's safe to do this, so perhaps a tiny bit more detail here?